src/app/payments/get-protocol-fee.ts

```diff
@@ -0,0 +1,129 @@
+import { NewDealerPriceService } from "@services/dealer-price"
+
+import {
+  constructPaymentFlowBuilder,
+  newCheckIntraledgerLimits,
+  newCheckWithdrawalLimits,
+} from "./helpers"
+
+export const getLightningFeeEstimation = async ({
+  walletId,
+  paymentRequest,
+}: {
+  walletId: string
+  paymentRequest: EncodedPaymentRequest
+}): Promise<PaymentAmount<WalletCurrency> | ApplicationError> => {
+  const decodedInvoice = decodeInvoice(paymentRequest)
+  if (decodedInvoice instanceof Error) return decodedInvoice
+  if (decodedInvoice.paymentAmount === null) {
+    return new LnPaymentRequestNonZeroAmountRequiredError()
+  }
+
+  return estimateLightningFee({
+    uncheckedSenderWalletId: walletId,
+    invoice: decodedInvoice,
+  })
+}
+
+export const getNoAmountLightningFeeEstimation = async ({
+  walletId,
+  paymentRequest,
+  amount,
+}: {
+  walletId: string
+  paymentRequest: EncodedPaymentRequest
+  amount: number
+}): Promise<PaymentAmount<WalletCurrency> | ApplicationError> => {
+  const decodedInvoice = decodeInvoice(paymentRequest)
+  if (decodedInvoice instanceof Error) return decodedInvoice
+
+  const { amount: lnInvoiceAmount } = decodedInvoice
+  if (lnInvoiceAmount && lnInvoiceAmount > 0) {
+    return new LnPaymentRequestZeroAmountRequiredError()
+  }
+
+  return estimateLightningFee({
+    uncheckedSenderWalletId: walletId,
+    invoice: decodedInvoice,
+    uncheckedAmount: amount,
+  })
+}
+
+const estimateLightningFee = async ({
+  uncheckedSenderWalletId,
+  invoice,
+  uncheckedAmount,
+}: {
+  uncheckedSenderWalletId: string
+  invoice: LnInvoice
+  uncheckedAmount?: number
+}): Promise<PaymentAmount<WalletCurrency> | ApplicationError> => {
+  const senderWalletId = checkedToWalletId(uncheckedSenderWalletId)
+  if (senderWalletId instanceof Error) return senderWalletId
+
+  const senderWallet = await WalletsRepository().findById(senderWalletId)
+  if (senderWallet instanceof Error) return senderWallet
+
+  const dealer = NewDealerPriceService()
+  const builder = await constructPaymentFlowBuilder({
+    senderWallet,
+    invoice,
+    uncheckedAmount,
+    usdFromBtc: dealer.getCentsFromSatsForImmediateBuy,
+    btcFromUsd: dealer.getSatsFromCentsForImmediateSell,
+  })
+  if (builder instanceof Error) return builder
+
+  const usdPaymentAmount = await builder.usdPaymentAmount()
+  if (usdPaymentAmount instanceof Error) return usdPaymentAmount
+  const btcPaymentAmount = await builder.btcPaymentAmount()
+  if (btcPaymentAmount instanceof Error) return btcPaymentAmount
+
+  const priceRatio = PriceRatio({ usd: usdPaymentAmount, btc: btcPaymentAmount })
+  if (priceRatio instanceof Error) return priceRatio
+
+  let paymentFlow
+  if (!(await builder.needsRoute())) {
+    const limitCheck = await newCheckIntraledgerLimits({
+      amount: usdPaymentAmount,
+      wallet: senderWallet,
+      priceRatio,
+    })
+    if (limitCheck instanceof Error) return limitCheck
+
```

Comment:
this should be prior to the `if` to avoid duplication?

Reply:
If you mean the limitCheck calc, it's two different functions (`IntraledgerLimits` vs `WithdrawalLimits`)